mdata/notifierKafka/notifierKafka.go

```diff
@@ -4,202 +4,85 @@ import (
-	c.wg.Add(1)
-	defer c.wg.Done()
-
-	pc, err := c.consumer.ConsumePartition(topic, partition, currentOffset)
+	err = c.consumer.StartAndAwaitBacklog(backlogProcessTimeout)
 	if err != nil {
-		log.Fatal(4, "kafka-cluster: failed to start partitionConsumer for %s:%d. %s", topic, partition, err)
+		log.Fatal(4, "kafka-cluster: Failed to start consumer: %s", err)
 	}
-	log.Info("kafka-cluster: consuming from %s:%d from offset %d", topic, partition, currentOffset)
 
-	messages := pc.Messages()
-	ticker := time.NewTicker(offsetCommitInterval)
-	startingUp := true
-	// the bootTimeOffset is the next available offset. There may not be a message with that
-	// offset yet, so we subtract 1 to get the highest offset that we can fetch.
-	bootTimeOffset := bootTimeOffsets[partition] - 1
-	partitionOffsetMetric := partitionOffset[partition]
-	partitionLogSizeMetric := partitionLogSize[partition]
-	partitionLagMetric := partitionLag[partition]
-	for {
-		select {
-		case msg := <-messages:
-			if mdata.LogLevel < 2 {
-				log.Debug("kafka-cluster received message: Topic %s, Partition: %d, Offset: %d, Key: %x", msg.Topic, msg.Partition, msg.Offset, msg.Key)
-			}
-			mdata.Handle(c.metrics, msg.Value, c.idx)
-			currentOffset = msg.Offset
-		case <-ticker.C:
-			if err := c.offsetMgr.Commit(topic, partition, currentOffset); err != nil {
-				log.Error(3, "kafka-cluster failed to commit offset for %s:%d, %s", topic, partition, err)
-			}
-			if startingUp && currentOffset >= bootTimeOffset {
-				processBacklog.Done()
-				startingUp = false
-			}
-			offset, err := c.client.GetOffset(topic, partition, sarama.OffsetNewest)
-			if err != nil {
-				log.Error(3, "kafka-mdm failed to get log-size of partition %s:%d. %s", topic, partition, err)
-			} else {
-				partitionLogSizeMetric.Set(int(offset))
-			}
-			if currentOffset < 0 {
-				// we have not yet consumed any messages.
-				continue
-			}
-			partitionOffsetMetric.Set(int(currentOffset))
-			if err == nil {
-				partitionLagMetric.Set(int(offset - currentOffset))
-			}
-		case <-c.stopConsuming:
-			pc.Close()
-			if err := c.offsetMgr.Commit(topic, partition, currentOffset); err != nil {
-				log.Error(3, "kafka-cluster failed to commit offset for %s:%d, %s", topic, partition, err)
-			}
-			log.Info("kafka-cluster consumer for %s:%d ended.", topic, partition)
-			return
-		}
-	}
+	go c.produce()
+
+	return &c
+}
+
+func (c *NotifierKafka) handleMessage(data []byte, partition int32) {
+	mdata.Handle(c.metrics, data, c.idx)
 }
 
-// Stop will initiate a graceful stop of the Consumer (permanent)
-//
-// NOTE: receive on StopChan to block until this process completes
 func (c *NotifierKafka) Stop() {
-	// closes notifications and messages channels, amongst others
-	close(c.stopConsuming)
+	log.Info("kafka-notifier: stopping kafka input")
 	c.producer.Close()
-
-	go func() {
-		c.wg.Wait()
-		c.offsetMgr.Close()
-		close(c.StopChan)
-	}()
+	c.consumer.Stop()
+	close(c.stopChan)
```

Comment:
isn't this racey? in between producer.Close and close stopChan, the produce goroutine may trigger a flush on the producer

Reply:
related: it would be a bit cleaner to flush any pending stuff before shutting down

Reply:
not sure if i'm totally missing something... but where is the `Stop()` on the notifiers actually called? I can't find it